lib/internal/modules/esm/resolve.js

```diff
@@ -153,44 +160,29 @@ function fileExists(url) {
  * @returns {URL}
  */
 function legacyMainResolve(packageJSONUrl, packageConfig, base) {
-  let guess;
-  if (packageConfig.main !== undefined) {
-    // Note: fs check redundances will be handled by Descriptor cache here.
-    if (fileExists(guess = new URL(`./${packageConfig.main}`,
-                                   packageJSONUrl))) {
-      return guess;
-    } else if (fileExists(guess = new URL(`./${packageConfig.main}.js`,
-                                          packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}.node`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.js`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.json`,
-                                        packageJSONUrl)));
-    else if (fileExists(guess = new URL(`./${packageConfig.main}/index.node`,
-                                        packageJSONUrl)));
-    else guess = undefined;
-    if (guess) {
-      emitLegacyIndexDeprecation(guess, packageJSONUrl, base,
-                                 packageConfig.main);
-      return guess;
+  const resolvedOption = URLLegacyMainResolve(packageJSONUrl.href, packageConfig.main);
+
+  // less than zero means that happened some error
+  // -2: LegacyMainResolveReturnType::kInvalidUrl
+  // -1: LegacyMainResolveReturnType::kModuleNotFound
+  if (resolvedOption < 0) {
+    if (resolvedOption === -2) {
+      throw new ERR_INVALID_URL(packageConfig.main || './index.js');
```

Comment:
Is it a new exception for this path? I don't see it in the previous version.

More specifically, are you sure it's not a breaking change?

Reply:
It's not a new exception, this if statement represents this line:

https://github.com/nodejs/node/blob/5c27cc2afbe77d4978a7c773c1dcf2e738aad7e7/lib/internal/modules/esm/resolve.js#L159-L160

There are two cases that the URL can be invalid, when `packageConfig.main` represents an invalid URL or when the `packageJSONUrl` cannot be parsed as the base URL.

Reply:
It's not a breaking change. `new URL` can throw `ERR_INVALID_URL`